src/permission-set.js

```diff
@@ -16,523 +16,809 @@
+    if (this.isAcl(resourceUri)) {
+      return resourceUri  // .acl resources are their own ACLs
+    } else {
+      return resourceUri + this.aclSuffix
+    }
+  }
+
   /**
-   * The URL of the corresponding ACL resource, at which these permissions will
-   * be saved.
-   * @property aclUrl
-   * @type {String}
+   * Tests whether a given uri is for an ACL resource.
+   * This is the default `isAcl()` method that's used by
+   * PermissionSet instances, unless it's overridden in options.
+   * @method defaultIsAcl
+   * @param uri {String}
+   * @returns {Boolean}
    */
-  this.aclUrl = aclUrl
+  static defaultIsAcl (uri) {
+    return uri.endsWith(this.aclSuffix)
+  }
+
   /**
-   * RDF Library (optionally injected)
-   * @property rdf
-   * @type {RDF}
+   * Adds a given Authorization instance to the permission set.
+   * Low-level function, clients should use `addPermission()` instead, in most
+   * cases.
+   * @method addAuthorization
+   * @private
+   * @param auth {Authorization}
+   * @return {PermissionSet} Returns self (chainable)
    */
-  this.rdf = options.rdf
+  addAuthorization (auth) {
+    var hashFragment = auth.hashFragment()
+    if (hashFragment in this.authorizations) {
+      // An authorization for this agent and resource combination already exists
+      // Merge the incoming access modes with its existing ones
+      this.authorizations[ hashFragment ].mergeWith(auth)
+    } else {
+      this.authorizations[ hashFragment ] = auth
+    }
+    if (!auth.virtual && auth.allowsControl()) {
+      // If acl:Control is involved, ensure implicit rules for the .acl resource
+      this.addControlPermissionsFor(auth)
+    }
+    // Create the appropriate indexes
+    this.addToAgentIndex(auth.webId(), auth.accessType, auth.resourceUrl, auth)
+    if (auth.isPublic()) {
+      this.addToPublicIndex(auth.resourceUrl, auth.accessType, auth)
+    }
+    return this
+  }
+
   /**
-   * Whether this permission set is for a 'container' or a 'resource'.
-   * Determines whether or not the inherit/'acl:default' attribute is set on
-   * all its Authorizations.
-   * @property resourceType
-   * @type {String}
+   * Creates an Authorization with the given parameters, and passes it on to
+   * `addAuthorization()` to be added to this PermissionSet.
+   * Essentially a convenience factory method.
+   * @method addAuthorizationFor
+   * @private
+   * @param resourceUrl {String}
+   * @param inherit {Boolean}
+   * @param agent {String|Quad} Agent URL (or `acl:agent` RDF triple).
+   * @param accessModes {String} 'READ'/'WRITE' etc.
+   * @param [origins=[]] {Array<String>} List of origins that are allowed access
+   * @param [mailTos=[]] {Array<String>}
+   * @return {Authorization}
    */
-  this.resourceType = isContainer ? CONTAINER : RESOURCE
+  addAuthorizationFor (resourceUrl, inherit, agent, accessModes, origins = [],
+                       mailTos = []) {
+    let auth = new Authorization(resourceUrl, inherit)
```

Comment:
They were initially being set in the constructor. I pulled em out of it because setting the agent/mode etc required additional logic. And this was written using es5 syntax, so that setters weren't convenient. But even in es6, I think I'd prefer them as separate methods.